patterns/interpreter.md

```diff
@@ -0,0 +1,185 @@
+language](https://en.wikipedia.org/wiki/Domain-specific_language), then define
+a grammar for this language and design interpreter solving problem
+instances.
+
+## Motivation
+
+Our goal is translate simple mathematical expressions into
+postfix expressions (or [Reverse Polish notation](https://en.wikipedia.org/wiki/Reverse_Polish_notation))
+For simplicity, our expressions consists of ten digits `0`,...,`9` and two
+operations `+, -`. For example, the expression `2 + 4` is translated into `2 4  +`.
+
+## Context Free Grammars
+
+In order to understand the Interpreter pattern you need to know a little bit
+about formal language theory. If you already know what a grammar is feel free
+to skip this section, otherwise fasten your seat belts.
+
+Basicaly, a [Context Free Grammar (CFG)](https://en.wikipedia.org/wiki/Context-free_grammar)
+defines a set of rules that describe all possible strings over some finite
+set of symbols, formally called alphabet. These strings usually have some well
+defined structure. For example, strings of all balanced parentheses:
+
+```ignore
+(), (()), ()(), ()(()), ..
+```
+
+Formally a CFG consists of terminal symbols, non-terminal symbols, a start
+(non-terminal) symbol, and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`. We define a
+single non-terminal (at the same time the start) symbol `S` and production rules
+as:
+
+```ignore
+S -> SS
+S -> (S)
+S -> ()
+```
+
+So, in order to generate a string with balanced parentheses we repeatedly apply
+production rules. For example, to derive `()()` we apply the following rules:
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+## Context Free Grammar for our problem
+
+Recall that our task is translate infix expressions into postfix ones. Let's
+define a context free grammar for a set of infix expressions over `0,...,9, +,-`,
+where:
+
+- terminal symbols: `0,...,9, +,-`
+- non-terminal symbols: `exp, term`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed depending on what we are going
+to do with the grammar. For example, we might need to remove left recursion.
+For more details please see [Compilers:Principles,Techniques, and Tools](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution
+
+We simply implement a recursive descent parser. The code panics when an expression
+is syntactically wrong (for exmple `2-34` or `2+5-` are wrong according to the
+grammar definition).
+
+```rust
+pub struct Interpreter<'a> {
+    it: std::str::Chars<'a>,
+}
+
+impl<'a> Interpreter<'a> {
+
+    pub fn new(infix: &'a String) -> Self {
+        Self { it: infix.chars() }
+    }
+
+    pub fn interpret(&mut self, out: &mut String) {
+        self.exp(out);
+    }
```

Comment:
what's the point of having this function? 🤔
It's the same as `exp`

Reply:
@MarcoIeni  I just follow the convention. Trait or struct must implement the function `interpret `, though I don't know why it is required. Look [https://en.wikipedia.org/wiki/Interpreter_pattern](url) or GoF book.

Reply:
Ok, if you like it you could delete this function and rename `exp` into `interpret` (and make it pub).
Otherwise if you like things as they are now just mark this conversation as resolved :)